content/courses/advanced-ada/parts/data_types/records.rst

```diff
@@ -2918,20 +2928,684 @@ In this example, :ada:`Arr_Sub` and :ada:`Rec_Sub` are unconstrained subtypes.
+
+.. code-block:: ada
+
+       type Device (State : Device_State) is
+       record
+          case State is
+             when Off =>
+                null;
+             when On =>
+                Info : Device_Info;
+          end case;
+       end record;
+
+For the sake of this example, we could say that a device that is turned off
+doesn't have any valuable information. Therefore, the device information stored
+in the :ada:`Info` component of the :ada:`Device` type is only valid if the
+device is turned on. Thus, if the device is turned off (i.e.,
+:ada:`Device_State = Off`), we should prevent the application from processing
+device information that is probably incorrect. Let's extend the previous code
+example to accommodate this requirement:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device
+
+    package Devices is
+
+       type Device_State is
+         (Off, On);
+
+       type Device
+         (State : Device_State := Off) is
+           private;
+
+       procedure Turn_Off (D : in out Device);
+
+       procedure Turn_On (D : in out Device);
+
+       type Device_Info is
+       record
+          V : Float;
+       end record;
+
+       function Read_Info (D : Device)
+                           return Device_Info;
+
+    private
+
+       type Device (State : Device_State := Off) is
+       record
+          case State is
+             when Off =>
+                null;
+             when On =>
+                Info : Device_Info;
+          end case;
+       end record;
+
+       Device_Off : constant Device :=
+                      (State => Off);
+
+       Device_On  : constant Device :=
+                      (State => On,
+                       others => <>);
+
+    end Devices;
+
+    package body Devices is
+
+       procedure Turn_Off (D : in out Device) is
+       begin
+          D := Device_Off;
+       end Turn_Off;
+
+       procedure Turn_On (D : in out Device) is
+       begin
+          D := Device_On;
+       end Turn_On;
+
+       function Read_Info (D : Device)
+                           return Device_Info is
+         (D.Info);
+
+    end Devices;
+
+Let's create a test application called :ada:`Show_Device` that makes use of
+this device by turning it on and off, and by retrieving information from it:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device
+    :class: ada-run-expect-failure
+
+    with Devices; use Devices;
+
+    procedure Show_Device is
+       D : Device;
+       I : Device_Info;
+    begin
+       Turn_On (D);
+       I := Read_Info (D);
+
+       Turn_Off (D);
+
+       --  The following call raises
+       --  an exception at runtime
+       --  because D is turned off.
+       I := Read_Info (D);
+    end Show_Device;
+
+In this example, by using the variant part, we're preventing that information
+retrieved by an inappropriate call to the :ada:`Read_Info` function is used
```

Comment:
The phrase "we're preventing that information retrieved" is a bit awkward in English.
Maybe  "we're preventing the case in which information retrieved"  instead?
Or
"we're preventing information retrieved by an inappropriate call to the :ada:`Read_Info` function from being used"

Reply:
Yes, that sounds indeed better.